(VDB-1774) Add Sentry error reporting

rmulhol opened the pull request:
- Use logrus hook to forward anything that's logged at a Fatal, Panic,
  or Error level.
- Requires configuring Sentry DSN and Sentry environment, but still
  runs without error if these are not configured.

Commit:
(VDB-1774) Add Sentry error reporting

- Use logrus hook to forward anything that's logged at a Fatal, Panic,
  or Error level.
- Requires configuring Sentry DSN and Sentry environment, but still
  runs without error if these are not configured.

rmulhol commented on cmd/root.go:
```diff
@@ -102,6 +107,34 @@ func logLevel() error {
 
+func setupSentryHook() error {
+	sentryEnv := viper.GetString("sentry.env")
+	sentryDSN := viper.GetString("sentry.dsn")
```
maybe it makes sense to return early if these aren't assigned? Have verified everything runs without them, but we do see stuff like this in the logs:
```
[Sentry] 2021/01/18 10:20:00 Sentry client initialized with an empty DSN. Using noopTransport. No events will be delivered.
[Sentry] 2021/01/18 10:20:00 Integration installed: ContextifyFrames
[Sentry] 2021/01/18 10:20:00 Integration installed: Environment
[Sentry] 2021/01/18 10:20:00 Integration installed: Modules
[Sentry] 2021/01/18 10:20:00 Integration installed: IgnoreErrors
```
I'll bet that would go away if we weren't initializing Sentry and adding a hook.

ecmendenhall replied:
Yeah, probably not huge overhead here but no harm in bailing early if the client's not using Sentry.

rmulhol commented on cmd/root.go:
```diff
@@ -102,6 +107,34 @@ func logLevel() error {
+	sentryErr := sentry.Init(sentry.ClientOptions{
+		Dsn:         sentryDSN,
+		Environment: sentryEnv,
+		Debug:       true,
```
Assuming I don't want to leave this as set to true as we merge to prod? Should it be made configurable?

ecmendenhall replied:
Definitely disable this in prod. I don't have a strong opinion on exposing this as a config option—I'd probably be OK removing it altogether.

rmulhol commented on cmd/root.go:
```diff
@@ -102,6 +107,34 @@ func logLevel() error {
+	sentryHook, hookErr := logrus_sentry.NewSentryHook(sentryDSN, []logrus.Level{
+		logrus.ErrorLevel,
+		logrus.FatalLevel,
+		logrus.PanicLevel,
```
One interesting thing here is that I'm not sure we log around all of our `panic`s - assuming we may need to go add that to make this effective?

ecmendenhall replied:
Yes, I suspect so, or handle them [separately](https://docs.sentry.io/platforms/go/panics/).

ecmendenhall replied:
(I'm totally in favor of merging this and doing that in a separate pass, though)

rmulhol replied:
Right now I'm leaning toward handing this in a separate story, because it seems like a lot of the built in approaches won't play nice with a `panic` emanating from arbitrary goroutines: https://github.com/getsentry/sentry-go/issues/115#issuecomment-564032753

That said I could totally be reading something wrong and am on board to revisit within this PR.

Sidebar is that I don't think we have that many `panic` statements and could probably replace all of them without too much effort.

rmulhol commented:
I believe this also resolves VDB-1775 and VDB-1776 since the changes to `root.go` will apply to every command

rmulhol commented on cmd/root.go:
```diff
@@ -24,6 +24,8 @@ import (
 
 	"github.com/ethereum/go-ethereum/ethclient"
 	"github.com/ethereum/go-ethereum/rpc"
+	"github.com/evalphobia/logrus_sentry"
```
Found this at https://github.com/sirupsen/logrus/wiki/Hooks but I'd be fine with hand rolling our own if we don't want to depend on a repo we don't own for this behavior

ecmendenhall commented:
I noticed the Sentry errors this hook is sending don't have attached stacktraces. That's a configuration option in the hook we can enable: https://github.com/evalphobia/logrus_sentry#enabling-stacktraces

Commit:
Address PR feedback

- Remove debug mode
- Add error stacktrace
- Bail Sentry setup if not configured
